### PR TITLE
feat(generator): allow excluding exported routes using `export:page` hook

### LINKED ISSUE #7440: Ability to prevents page generation in `generate:page`
### What problem does this feature solve?

What if we can add a single boolean to the `page` variable to conditionally control page generation?

https://github.com/nuxt/nuxt.js/blob/c14471681c7134a6a08a69f9eabdb7d213feb2f0/packages/generator/src/generator.js#L361-L373

### What does the proposed changes look like?

```js
// Call hook to let user update the path & html
    const page = { route, path: fileName, html, generate: true }
    await this.nuxt.callHook('generate:page', { page, errors: pageErrors })
    await this.nuxt.callHook('export:page', { page })
  
    if (page.generate) {
        page.path = path.join(this.distPath, page.path)
        ...
    }

    return page.generate

```

Or for being fully backward compatible we could use `export:page` hook:
```js
    const page = { route, path: fileName, html, generate: true }
...
    await this.nuxt.callHook('export:page', { page, errors: pageErrors })
  
    if (page.generate) {
...
```

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c10713">#c10713</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by farnabaz**
close #7440

### CODE REVIEW

File: test/dev/basic.generate.test.js [hunk 1]
```diff
@@ -36,6 +36,13 @@ describe('basic generate', () => {
     nuxt.hook('generate:done', () => {
       writeFileSync(changedFileName, '')
     })
+    nuxt.hook('export:page', ({ page, errors }) => {
+      if (errors.length > 0) {
+        if (page.path.indexOf('/skip-on-fail') === 0) {
```
Comment (pi0) [inline test/dev/basic.generate.test.js:41]: Refactor tips:
 - `> 0` can be omitted
- `indexOf === 0` can be `includes` (or simply eqeqeq)
- Both conditions can be combined

File: packages/generator/src/generator.js [hunk 2]
```diff
@@ -359,27 +359,29 @@ export default class Generator {
...
 
     // Call hook to let user update the path & html
-    const page = { route, path: fileName, html }
+    const page = { route, path: fileName, html, export: true }
     await this.nuxt.callHook('generate:page', page)
-    await this.nuxt.callHook('export:page', { page })
+    await this.nuxt.callHook('export:page', { page, errors: pageErrors })
 
-    page.path = path.join(this.distPath, page.path)
+    if (page.export) {
```
Comment (pi0) [inline packages/generator/src/generator.js:366]: We can early return here like `if (!page.export) { return false }`

File: packages/generator/src/generator.js [hunk 3]
```diff
@@ -359,27 +359,29 @@ export default class Generator {
     }
 
     // Call hook to let user update the path & html
-    const page = { route, path: fileName, html }
+    const page = { route, path: fileName, html, export: true }
```
Comment (pi0) [inline packages/generator/src/generator.js:362]: Maybe `skip: false` makes more sense?

Comment (farnabaz) [reply packages/generator/src/generator.js]: I will go with `exclude` :)

**Review by atinux — DISMISSED** [review]

**Comment by Atinux** [comment]
Approved but missed the failing checks

**Review by pi0 — APPROVED** [review]
Great!

**Pull request merged**